dawn/src/dawn/AST/Offsets.h

```diff
@@ -71,25 +71,42 @@ class CartesianOffset : public HorizontalOffsetImpl {
 
 class HorizontalOffset {
 public:
+  HorizontalOffset() = default;
+
   explicit HorizontalOffset(cartesian_) : impl_(std::make_unique<CartesianOffset>()) {}
   HorizontalOffset(cartesian_, int iOffset, int jOffset)
       : impl_(std::make_unique<CartesianOffset>(iOffset, jOffset)) {}
   HorizontalOffset(HorizontalOffset const& other) { *this = other; }
   HorizontalOffset& operator=(HorizontalOffset const& other) {
-    impl_ = other.impl_->clone();
+    if(other.impl_)
+      impl_ = other.impl_->clone();
+    else
+      impl_ = nullptr;
     return *this;
   }
   HorizontalOffset(HorizontalOffset&& other) = default;
   HorizontalOffset& operator=(HorizontalOffset&& other) = default;
 
-  bool operator==(HorizontalOffset const& other) const { return *impl_ == *other.impl_; }
+  bool operator==(HorizontalOffset const& other) const {
```

Comment:
Are you sure that you would expect that? I feel like comparing a cartesian and an unstructured offset is always an error, because we cannot and we will never be able to handle that case. Just returning false, would mean that they are somewhat compatible, but they are not, in my opinion.

Reply:
Mixing them should crash. What I meant was slightly unrelated: sometimes it may not be clear from the context whether we are dealing with an unstructured/cartesian stencil, therefore we would "ask" `HorizontalOffset` which is its type and run different code accordingly.

Reply:
To wrap up our discussion around this: we should wait if there's an actual usecase for this "getType" on an offset, but probably there shouldn't be any.